dashboard/test/controllers/admin_users_controller_test.rb

```diff
@@ -269,5 +269,57 @@ class AdminUsersControllerTest < ActionController::TestCase
     assert_select 'td', text: @facilitator.id.to_s
   end
 
+  test 'bulk_grant_permission grants multiple user_permissions' do
+    sign_in @admin
+    teacher = create(:teacher, username: 'teacher', email: 'teacher@email.xx')
+    assert_difference 'UserPermission.count', 2 do
+      post :bulk_grant_permission, params: {emails: "#{@not_admin.email}\r\n#{teacher.email}", bulk_permission: UserPermission::LEVELBUILDER}
+    end
+    assert_redirected_to permissions_form_path
+    assert @not_admin.reload.permission?(UserPermission::LEVELBUILDER), 'Permission not granted to user'
+    assert teacher.reload.permission?(UserPermission::LEVELBUILDER), 'Permission not granted to user'
+    assert_equal(
+      "#{UserPermission::LEVELBUILDER.titleize} Permission added for 2 Users",
+      flash[:notice]
+    )
+    assert_nil flash[:alert]
+  end
+
+  test 'bulk_grant_permission does not grant user_permissions for student user' do
+    sign_in @admin
+    student_email = 'student@email.xx'
+    student = create :student, email: student_email
+    assert_does_not_create UserPermission do
+      post :bulk_grant_permission, params: {emails: student_email, bulk_permission: UserPermission::LEVELBUILDER}
+    end
+    assert_redirected_to permissions_form_path
+    refute student.reload.permission?(UserPermission::LEVELBUILDER), 'Permission granted to student'
+    assert_equal(
+      "FAILED: These Users could not be found or are not teachers: #{student_email}",
+      flash[:alert]
+    )
+    assert_nil flash[:notice]
+  end
+
+  test 'bulk_grant_permission only grants user_permission for teachers in mixed list' do
+    sign_in @admin
+    student_email = 'student@email.xx'
+    student = create :student, email: student_email
+    assert_difference 'UserPermission.count' do
+      post :bulk_grant_permission, params: {emails: "#{@not_admin.email}\r\n#{student_email}", bulk_permission: UserPermission::LEVELBUILDER}
+    end
+    assert_redirected_to permissions_form_path
+    assert @not_admin.reload.permission?(UserPermission::LEVELBUILDER), 'Permission not granted to user'
+    refute student.reload.permission?(UserPermission::LEVELBUILDER), 'Permission granted to student'
+    assert_equal(
+      "#{UserPermission::LEVELBUILDER.titleize} Permission added for 1 User",
+      flash[:notice]
+    )
+    assert_equal(
+      "FAILED: These Users could not be found or are not teachers: #{student_email}",
+      flash[:alert]
+    )
+  end
```

Comment:
Good testing :)

It might be worth adding an additional test for idempotence (even though that's handled in `User.permission=`) - create a user with a given permission, then bulk assign and make sure it still has the permission and there are no errors.

Reply:
good point! I'll add that